drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,692 @@
+/*
+ * Copyright (C) ??? 2023 
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sx126x
+ * @{
+ * @file
+ * @brief       Radio HAL implementation for the Semtech SX126x 
+ *
+ * @author      Klim Evdokimov <klimevdokimov@mail.ru>
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#include "net/ieee802154/radio.h"
+#include "event/callback.h"
+#include "event/thread.h"
+#include "ztimer.h"
+#include "checksum/ucrc16.h"
+#include "kernel_defines.h"
+#include "event.h"
+
+#include "sx126x.h"
+#include "sx126x_params.h"
+#include "sx126x_internal.h"
+
+#define LORA_ACK_REPLY_US          1024
+
+#ifndef SX126X_HAL_CHAN_BASE
+#define SX126X_HAL_CHAN_BASE (868300000LU)
+#endif
+
+#ifndef SX126X_HAL_CHAN_SPACING
+#define SX126X_HAL_CHAN_SPACING (200000LU)
+#endif
+
+#define SX126X_CHAN_MIN (IEEE802154_CHANNEL_MIN_SUBGHZ)
+
+#ifndef SX126X_CHAN_MAX
+#define SX126X_CHAN_MAX (8)
+#endif
+
+#define SX126X_POWER_MIN (-9)
+
+#ifndef SX126X_POWER_MAX
+#define SX126X_POWER_MAX (22)
+#endif
+
+
+static const ieee802154_radio_ops_t sx126x_ops;
+
+static int _set_state(sx126x_t *dev, sx126x_state_t state);
+static int _get_state(sx126x_t *dev, void* val);
+
+void _sx126x_handler(void* arg)
+{
+#if IS_USED(MODULE_IEEE802154)
+    ieee802154_dev_t *hal = arg;
+    sx126x_hal_task_handler(hal);
+#else
+    netdev_t *dev = arg;
+    netdev_trigger_event_isr(dev);
+    
+#endif
+}
+
+#if IS_USED(MODULE_SX126X_STM32WL)
+
+event_callback_t sx126x_ev_callback;
+#if IS_USED(MODULE_IEEE802154)
+void isr_subghz_radio(void)
+{
+    /* Disable NVIC to avoid ISR conflict in CPU. */
+    NVIC_DisableIRQ(SUBGHZ_Radio_IRQn);
+    NVIC_ClearPendingIRQ(SUBGHZ_Radio_IRQn);
+    event_post(EVENT_PRIO_MEDIUM, (event_t*)&sx126x_ev_callback);
```

Comment:
I just realized this is being posted to a queue running in a thread that has different priority than the thread that runs the driver (`gnrc_netif_ieee802154`).
To solve that, you should probably set an `event_queue_t` pointer (e.g in `sx126x_init`) and do:
```suggestion
    event_post(sx126x_dev->evq, (event_t*)&sx126x_ev_callback);
```

This `sx126x_dev` variable should point to the (unique) device descriptor (available only when using the `perioph` version of the driver)

Reply:
Hello @jia200x 

Sorry for the delay. 
I changed priority of the event thread by `USEMODULE += event_thread_highest` but it didn't solve the problem with acks. Probably I don't understand the way that you suggest. I just don't know how I should put the hal pointer from ISR handler except the event_callback mechanism. And I have not `periph` version of this module to check this idea with queue :(